<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Проект-SQL" data-toc-modified-id="Проект-SQL-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Проект SQL</a></span></li><li><span><a href="#Посчитайте,-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитайте,-сколько-книг-вышло-после-1-января-2000-года-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Посчитайте, сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку;" data-toc-modified-id="Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку;-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Для каждой книги посчитайте количество обзоров и среднюю оценку;</a></span></li><li><span><a href="#Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще50-страниц-—-так-вы-исключите-из-анализа-брошюры" data-toc-modified-id="Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще50-страниц-—-так-вы-исключите-из-анализа-брошюры-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Определите издательство, которое выпустило наибольшее число книг толще50 страниц — так вы исключите из анализа брошюры</a></span></li><li><span><a href="#Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками" data-toc-modified-id="Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками</a></span></li><li><span><a href="#Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Общий вывод:</a></span></li></ul></div>

## Проект SQL



Коронавирус застал мир врасплох, изменив привычный порядок вещей. В
свободное время жители городов больше не выходят на улицу, не посещают кафе
и торговые центры. Зато стало больше времени для книг. Это заметили
стартаперы — и бросились создавать приложения для тех, кто любит читать.
Ваша компания решила быть на волне и купила крупный сервис для чтения книг
по подписке. Ваша первая задача как аналитика — проанализировать базу
данных.

В ней — информация о книгах, издательствах, авторах, а также пользовательские
обзоры книг. Эти данные помогут сформулировать ценностное предложение для
нового продукта.


**Спецификация**

Таблица books - содержит данные о книгах:

 - book_id — идентификатор книги;
 - author_id — идентификатор автора;
 - title — название книги;
 - num_pages — количество страниц;
 - publication_date — дата публикации книги;
 - publisher_id — идентификатор издателя.

Таблица authors- содержит данные об авторах:
 - author_id — идентификатор автора;
 - author — имя автора.

Таблица publishers -содержит данные об издательствах:
 - publisher_id — идентификатор издательства;
 - publisher — название издательства;

Таблица ratings - содержит данные о пользовательских оценках книг:
 - rating_id — идентификатор оценки;
 - book_id — идентификатор книги;
 - username — имя пользователя, оставившего оценку;
 - rating — оценка книги.

Таблица reviews -содержит данные о пользовательских обзорах на книги:
 - review_id — идентификатор обзора;
 - book_id — идентификатор книги;
 - username — имя пользователя, написавшего обзор;
 - text — текст обзора
 

**Цель исследования**
Необходимо ответить на следующие вопросы 
  - Посчитайте, сколько книг вышло после 1 января 2000 года;
  - Для каждой книги посчитайте количество обзоров и среднюю оценку;
  - Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
  - Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
  - Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [1]:
# импортируем библиотеки
import pandas as pd 
from sqlalchemy import create_engine 

In [2]:
# устанавливаем параметры
db_config = {'user': '******', # имя пользователя
             'pwd': '*****', # пароль
             'host': '******             'port': 6432, # порт подключения
             'db': '********'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], 
                                                         db_config['pwd'], 
                                                         db_config['host'], 
                                                         db_config['port'], 
                                                         db_config['db']) 
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
# получаем данные таблицы books
query = '''
            SELECT * FROM books
        '''
# подключения к базе данных для Sqlite
books = pd.io.sql.read_sql(query, con = engine)

books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
# проверим размерность и пропуски в таблице
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


В таблице 1000 строк , пропусков не выявлено

In [5]:
# получаем данные таблицы 
query = '''
            SELECT * FROM authors
        '''
# подключения к базе данных для Sqlite
authors = pd.io.sql.read_sql(query, con = engine)

authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
# проверим размерность и пропуски в таблице
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


В таблице 636 строки, пропусков не выявлено

In [7]:
# получаем данные таблицы 
query = '''
            SELECT * FROM publishers
        '''
# подключения к базе данных для Sqlite
publishers = pd.io.sql.read_sql(query, con = engine)

publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [8]:
# проверим размерность и пропуски в таблице
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


В таблице 340 строк , пропусков не выявлено

In [9]:
# получаем данные таблицы 
query = '''
            SELECT * FROM ratings
        '''
# подключения к базе данных для Sqlite
rating = pd.io.sql.read_sql(query, con = engine)

rating.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

В таблице 340 строк , пропусков не выявлено

In [10]:
# проверим размерность и пропуски в таблице
rating.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


В таблице 6456 строк , пропусков не выявлено

In [11]:
#  получаем данные таблицы 

query = '''
        SELECT * FROM reviews

        '''

# подключение к базе данных для sqlite
review= pd.io.sql.read_sql( query, con = engine)

In [12]:
# проверим размерность и пропуски в таблице
review.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


В таблице 2793 строк, пропусков не выявлено

In [13]:
def table (df):
    '''Функция по основной информации, пропускам, дубликатам в таблице
    '''
    print ('\n Основная информация по таблице \n')
    print (df.info())
    print('Наличие дубликатов {} \n '.format(df.duplicated().sum()))
    print('Наличие пропусков в таблице \n')
    print ((df.isna().sum()))

In [ ]:
#код ревьюера
#tables = ['table_1', 'table_2', 'table_3']
#
#for table_name in tables:
#    query = f'SELECT * FROM {table_name}'
#    table(pd.io.sql.read_sql(query, con = engine))

In [14]:
table(books)


 Основная информация по таблице 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None
Наличие дубликатов 0 
 
Наличие пропусков в таблице 

book_id             0
author_id           0
title               0
num_pages           0
publication_date    0
publisher_id        0
dtype: int64


In [15]:
table(authors)


 Основная информация по таблице 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None
Наличие дубликатов 0 
 
Наличие пропусков в таблице 

author_id    0
author       0
dtype: int64


In [16]:
table(review)


 Основная информация по таблице 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
None
Наличие дубликатов 0 
 
Наличие пропусков в таблице 

review_id    0
book_id      0
username     0
text         0
dtype: int64


In [17]:
table(rating)


 Основная информация по таблице 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
None
Наличие дубликатов 0 
 
Наличие пропусков в таблице 

rating_id    0
book_id      0
username     0
rating       0
dtype: int64


In [18]:
table(publishers)


 Основная информация по таблице 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
None
Наличие дубликатов 0 
 
Наличие пропусков в таблице 

publisher_id    0
publisher       0
dtype: int64


## Посчитайте, сколько книг вышло после 1 января 2000 года


In [19]:
query = '''
        SELECT 
        COUNT(book_id) as cnt_book ----количество книг
        FROM books
        WHERE publication_date>='2000-01-01'
        '''
pd.read_sql(query, con = engine)

cnt_book
0       821

 821 книга была опубликована после 1 января

## Для каждой книги посчитайте количество обзоров и среднюю оценку;


In [20]:
query = '''
    SELECT b.book_id, b.title, ---- индификатор книги и название
    COUNT(DISTINCT r.review_id) as cnt_rev, ----количество обзоров
    AVG(rt.rating) as avg_rating
    FROM BOOKS b 
        LEFT JOIN ratings rt ON rt.book_id=b.book_id 
        LEFT JOIN reviews r ON r.book_id=rt.book_id
    GROUP BY b.book_id
    ORDER BY cnt_rev DESC
    '''
pd.read_sql(query, con = engine)

book_id                                              title  cnt_rev  \
0        948                            Twilight (Twilight  #1)        7   
1        963                                Water for Elephants        6   
2        734                                   The Glass Castle        6   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...        6   
4        695  The Curious Incident of the Dog in the Night-Time        6   
..       ...                                                ...      ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel        0   
996      808                            The Natural Way to Draw        0   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites        0   
998      221                          Essential Tales and Poems        0   
999      191  Disney's Beauty and the Beast (A Little Golden...        0   

     avg_rating  
0      3.662500  
1      3.977273  
2      4.206897  
3      4.414634  
4      4.081081  
..          ...  
995    3.666667  
996    3.000000  
997    5.000000  
998    4.000000  
999    4.000000  

[1000 rows x 4 columns]

Больше всего обзоров у книги Twilight  - 7 обзоров

## Определите издательство, которое выпустило наибольшее число книг толще50 страниц — так вы исключите из анализа брошюры

In [21]:
query = '''
        SELECT p.publisher, ---- название издательства 
        COUNT (b.book_id) as cnt_book --- количество книг 
        FROM books b 
            LEFT JOIN  publishers p ON b.publisher_id=p.publisher_id 
        WHERE b.num_pages>50   
        GROUP BY p.publisher
        ORDER BY cnt_book DESC
        LIMIT 1
    '''
pd.read_sql(query, con = engine)

publisher  cnt_book
0  Penguin Books        42

Больше всего книг у издательства Penguin Books

## Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [22]:
query  = '''
        SELECT a.author,             ---- название автора
        AVG(rt.rating) AS avg_rating ----cредний рейтинг
        FROM 
         books b
        LEFT JOIN authors a ON b.author_id = a.author_id
        LEFT JOIN ratings rt ON rt.book_id = b.book_id
        WHERE b.book_id IN (
            SELECT b.book_id FROM books b    
                LEFT JOIN ratings rt ON rt.book_id = b.book_id     
                     GROUP BY
                          b.book_id
                     HAVING 
                          COUNT(rt.rating_id) >= 50
                      )
                 GROUP BY
                    a.author  
                 ORDER BY 
                    avg_rating DESC
                 LIMIT 5    
       '''
pd.read_sql(query, con = engine)



author  avg_rating
0         J.K. Rowling/Mary GrandPré    4.287097
1  Markus Zusak/Cao Xuân Việt Khương    4.264151
2                     J.R.R. Tolkien    4.246914
3                  Louisa May Alcott    4.192308
4                       Rick Riordan    4.080645

J.K. Rowling/Mary GrandPré - автор с самой высокой средней оценкой 

## Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [23]:
query = '''
     SELECT SUM(a.cnt) / COUNT(a.username) AS avg_cnt_reviews
     FROM
        (SELECT r.username,
            COUNT(r.review_id) AS cnt ---- количество оценок 
                FROM
                    reviews r 
                WHERE r.username IN  (SELECT rt.username FROM ratings rt
                     LEFT JOIN books b ON b.book_id = rt.book_id
                     GROUP BY
                         rt.username
                     HAVING
                         COUNT(rt.rating_id) > 50)
                GROUP BY r.username) AS a

       '''
pd.read_sql(query, con = engine).round()

avg_cnt_reviews
0             24.0

 Среднее количество обзоров на пользователя с более чем 50 оценками - 24


## Общий вывод:

    Больше всего обзоров у книги Twilight - 7 обзоров
    Самый рейтинговый писатель - J.K. Rowling/Mary GrandPré.
    Лидирующее издательство по выпуску книг - Penguin Books
    Среднее количество обзоров на пользователя с более чем 50 оценками - 24

